Spanner/src/Connection/Grpc.php

```diff
@@ -1092,4 +1157,70 @@ private function getDatabaseAdminClient()
 
         return $this->databaseAdminClient;
     }
+
+    /**
+     * Allow lazy instantiation of the spanner client.
+     * If routing is enabled, a SpannerClient is returned for a specific instanceName
+     * with a binding in the connection of a specific endpoint.
+     * If the endpoint list is not requested, a request is first made for this instanceName.
+     *
+     * @param string $instanceName Optional. Full name of instance.
+     * @return SpannerClient
+     */
+    private function getSpannerClient($instanceName = null)
+    {
+        if ($this->enableResourceCaching) {
```

Comment:
Proposing to refactor `getSpannerClient` a bit.

- If `spannerClients[$instanceName]` is set => return it
- elseIf
    - `enableResourceCaching` is set to `true`
        - try to get instance endpointUri
            - if successful update `spannerClients[$instanceName]` with specific client and return it
            - if unsuccessful update `spannerClients[$instanceName]` with default client and return it
    - if `enableResourceCaching` is set to `false`
        - update `spannerClients[$instanceName]` with default client and return it

```php
    private function getSpannerClient($instanceName)
    {
        if (isset($this->spannerClients[$instanceName])) {
            return $this->spannerClients[$instanceName];
        } elseif ($this->enableResourceCaching) {
            $projectId = InstanceAdminClient::parseName($instanceName)['project'];
            try {
                $instanceInfo = $this->getInstance([
                    'projectId' => $projectId,
                    'name' => $instanceName,
                    'fieldMask' => ['endpointUris']
                ]);
                if (empty($instanceInfo['endpointUris'])) {
                    return $this->getDefaultSpannerClient($instanceName);
                } else {
                    $grpcConfig = $this->grpcConfig;
                    // for now disregarding multiple endpointUris
                    $grpcConfig['apiEndpoint'] = $instanceInfo['endpointUris'][0];
                    return $this->spannerClients[$instanceName] =
                        $this->constructGapic(SpannerClient::class, $grpcConfig);
                }
            } catch (ApiException $ex) {
                if (ApiStatus::PERMISSION_DENIED == $ex->getStatus()) {
                    trigger_error(
                        'To use resource caching please add "spanner.instances.get" permission ',
                        E_USER_WARNING
                    );
                } else {
                    throw $ex;
                }
            }
        } 
        
        return $this->getDefaultSpannerClient($instanceName);
    }

    private function getDefaultSpannerClient($instanceName)
    {
        return $this->spannerClients[$instanceName] = 
            $this->spannerClients ?: $this->constructGapic(SpannerClient::class, $this->grpcConfig);
    }
```
WDYT